CRM/Core/Error.php

```diff
@@ -680,7 +680,13 @@ protected static function generateLogFileName($prefix) {
         $hash = self::generateLogFileHash($config) . '.';
       }
       else {
-        $hash = '';
+        // If Config And Log Directory is under the webroot still generate the hash file.
+        if (stripos($config->configAndLogDir, $config->userSystem->getCiviSourceStorage()['path'])) {
+          $hash = self::generateLogFileHash($config) . '.';
+        }
+        else {
+          $hash = '';
+        }
```

Comment:
I think I see what you're going for -- i.e. adding another safety-latch to prevent people from shooting themselves in the foot. That's generally good form, but I think it's not worth the effort here. A few specific thoughts...

(1) __Implementation of the check__ This `stripos(...)` conditional isn't doing what the comment/description says. Compare these values from my local dmaster:

* `getCiviSourceStorage()['path']` ==> `"/Users/me/bknix/build/dmaster/web/sites/all/modules/civicrm/"`
* `configAndLogDir` ==> `"/Users/me/bknix/build/dmaster/web/sites/default/files/civicrm/ConfigAndLog/"`

Note that `configAndLogDir` is under the web root (`/Users/me/bknix/build/dmaster/web`), but it is not under the `getCiviSourceStorage()`.  Consequently, the condition yields a false-negative:

```
$ cv ev '$config = CRM_Core_Config::singleton(); return stripos($config->configAndLogDir, $config->userSystem->getCiviSourceStorage()["path"]);'

false
```

It would come closer to compare with `getPath("[cms.root]/")`. (This formulation also respects any system-specific-overrides that might be in `$civicrm_paths`.) 

Tangentially, props for thinking about case-insensitive filesystems (`stripos(...)`)... I never thought about that before... OTOH, case-sensitive filesystems would need `strpos()`. In either case, the test should be for `stripos(...) === 0`. For a major curve-ball, there are all the permutations of symlinks that can be setup. `CRM_Utils_File::isChildPath($parent, $child)` should be more robust than either `strpos()` or `stripos()`. 

Putting it together, I think this is the closest to what it's trying to test:

```
$ cv ev '$p = Civi::paths(); return CRM_Utils_File::isChildPath($p->getPath("[cms.root]/"), $p->getPath("[civicrm.log]/"));'

true
```

(*For good measure, maybe call `isChildPath()` twice - once with `$checkRealPath=FALSE` and once with `$checkRealPath=TRUE`.... Hrmmm... symlinks....*)

(2) __Is a heuristic check worthwhile?__ At the end of the day, even this formulation ("Does logdir live under CMS root?") is fundamentally a *heuristic* for determining if the folder is web-accessible. We can certainly find cases where the heuristic gets it wrong, e.g.

* Maybe the CMS root isn't actually the web root. The datadir/logdir might live outside the CMS root but still be under the webroot. Similarly, there can be configurations with aliases/rewrites where something is web-accessible regardless.
* Maybe the datadir/logdir lives under the CMS root... but has suitable defenses for the folder. In which case, it is valid to use more predictable names (and thereby simplify your log-rotation/log-shipping setup). Someone in this situation will be perplexed that their choice of `CIVICRM_LOG_HASH` is not respected.

Of course, sometimes you have to use a heuristic despite its limits... e.g. if the next-best alternative (instead of using a heuristic) is to break most deployments or make most deployments insecure, then yes... use a heuristic. But in this case...

The thing that sticks out to me is that #17192 still preserves the default behavior, and the default behavior errs on the side of security (enabling the hash code). The sysadmin has to take special/unusual measures to opt-out of the secure mode (i.e. edit `civicrm.settings.php`, search the comments, and set `CIVICRM_LOG_HASH` to `false`). There's literally warning right next to the option about this.

(Analogy: `CIVICRM_LOG_HASH` is not a footgun. It's an uncharged taser safely packed in a sealed container with red tape. If a sysadmin cuts the tape, opens the package, charges it up, and then shoots himself in the foot... that's on him.)